### ISSUE TITLE
Support the ? operator within a function

### ISSUE THREAD

**Issue by martell** [google/evcxr#91]
It seems that we currently have to use `.unwrap()` everywhere we would prefer to use `?`

Example

```
:dep reqwest = { version = "0.10", features = ["blocking"] }
use reqwest::*;
let body = reqwest::blocking::get("https://www.rust-lang.org/")?.text()?;
body
```

Here is the error output
```
let body = reqwest::blocking::get("https://www.rust-lang.org/")?.text()?;
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot use the `?` operator in a function that returns `*mut evcxr_internal_runtime::VariableStore`
the `?` operator can only be used in a function that returns `Result` or `Option` (or another type that implements `std::ops::Try`)

let body = reqwest::blocking::get("https://www.rust-lang.org/")?.text()?;
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ cannot use the `?` operator in a function that returns `*mut evcxr_internal_runtime::VariableStore`
the `?` operator can only be used in a function that returns `Result` or `Option` (or another type that implements `std::ops::Try`)
```

**Comment by davidlattimore** [comment]
Interesting idea. It'd probably be necessary to decide on a specific error type for the return value. I'm somewhat unsure of how that would work out, since it'd be necessary for either evcxr, or something else to define `Into` implementations for each error type that could be returned.

**Comment by martell** [comment]
That is exactly where I got stuck on this one, so I made an issue about it. Maybe it could be something like SomeGenericEvcxrErrorType where the only goal is to print out the details of the error as the return value, but I would need to do a bit more reading on the try macro and the `Into` trait with some generic magic? to see if this is even possible.

**Comment by davidlattimore** [comment]
It could also be tricky making sure that variables don't get lost when an error gets returned. All the variables are stored in a HashMap. Each time some code gets executed, all the variables are taken out of the HashMap and put into variables with appropriate types, then put back into the HashMap when execution completes. Panics generally cause variables to get lost (although there is a mode where they get preserved, but it slows down compilation, so is currently off by default). But for returning an error, it seems that the user expectation would be that variables wouldn't get lost. That means that evcxr probably would need to expand the ? operator to some custom code that puts all the variables back into the HashMap then prints out the error message. I guess that means that we might not need to actually have an error return type, Into implementations etc - but it does mean we need to mess with the users code quite a bit - ideally without any side effects that the user would notice.

**Comment by davidlattimore** [comment]
If you build from head, you can now use the question mark operator provided the error type implements Display. It also works in conjunction with async code. So you can do `.await?`. I'll probably do a release in the not too distant future, but if anyone wants to give it a go before then and report any problems, that'd be appreciated.

To try it out, you can either git clone the repository, `cargo build`, then run `./target/debug/evcxr` or if you prefer you can just do:

```sh
cargo install --force --git https://github.com/google/evcxr.git evcxr_repl
cargo install --force --git https://github.com/google/evcxr.git evcxr_jupyter
```

**Comment by davidlattimore** [comment]
I found one problem. Then a question mark operator encounters an error, the early return means that variables that were supposed to be defined by the code being executed won't be defined, however the host doesn't know that. For example, if you execute:

```rust
let v = "hello".parse::<i32>()?;
```

Then subsequent executions will panic, reporting that the variable `v` has gone missing.

Two options come to mind to fix this.

1. Allow the host to query the guest to determine definitively what variables are currently available.
2. Have the guest signal the host to let it know that an error occurred and that any variables that were in the process of being defined should be dropped.

The end effect for the user should be the same in both cases. Preexisting variables, provided they were neither moved by the code being executed, nor shadowed by a new variable with the same name should still be available after an error.

**Comment by martell** [comment]
@davidlattimore wow this is an amazing turn around, I was only just thinking about and playing with the idea and you had it implemented and merged already, really great stuff going on here with evcxr.

**Comment by davidlattimore** [comment]
Thanks again for the suggestion @martell , it's definitely much nicer using ? than unwrap everywhere. Even more so, since when you get an error, it just prints it rather than panicking and dropping all your variables. I've just released 0.5.0, which includes this feature.
